### ISSUE TITLE
StreamField and StreamBlock have confusing behaviour with `required` and `blank` properties

### ISSUE THREAD

**Issue by nimasmi** [wagtail/wagtail#4306]
### Issue Summary

The documentation doesn't correspond to the actual behaviour of an optional top-level StreamBlock, and in some circumstances the behaviour is confusing even without the conflicting documentation.

From http://docs.wagtail.io/en/v1.13.1/topics/streamfield.html#streamblock:

> StreamBlock accepts the following options as either keyword arguments or Meta properties:
>
> `required` (default: True)
> If true, at least one sub-block must be supplied. This is ignored when using the `StreamBlock` as the top-level block of a StreamField; in this case the StreamField’s `blank` property is respected instead.

In fact, setting `required=False` in the StreamBlock and `blank=True` on the StreamField are both necessary to produce an optional field that also looks optional.

### Steps to Reproduce

Configure the following models.py:

``` python
from wagtail.wagtailadmin.edit_handlers import StreamFieldPanel
from wagtail.wagtailcore.blocks import StreamBlock, TextBlock
from wagtail.wagtailcore.models import Page
from wagtail.wagtailcore.fields import StreamField

class TestModel(Page):
    basic = StreamField(StreamBlock([('text', TextBlock())]))
    nonrequired = StreamField(StreamBlock([('text', TextBlock())], required=False))
    blank = StreamField(StreamBlock([('text', TextBlock())]), blank=True)
    nonrequired_blank = StreamField(StreamBlock([('text', TextBlock())], required=False), blank=True)

    content_panels = Page.content_panels + [
        StreamFieldPanel('basic'),
        StreamFieldPanel('nonrequired'),
        StreamFieldPanel('blank'),
        StreamFieldPanel('nonrequired_blank'),
    ]
```

Based on the documentation…

- `basic` should be a required field, which it is
- `nonrequired` should not be required*, and it is not (you can submit an empty field), but the red 'required' asterisk **is** shown
- `blank` should be not required _(in this case the StreamField’s `blank` property is respected instead)_, and yet it is, although the red 'required' asterisk **is not** shown
- `nonrequired_blank` should produce a blank block, but allegedly specifying `required=False` is superfluous, however this is the only way to get a non-required block that also looks optional to site editors.

\* though you'd think that a field with `blank=False` implicitly set would fail validation

### Technical details

* Wagtail version: 1.13.1

**Comment by coredumperror** [comment]
I'm suffering from this same problem. The following code results in a SteamBlock that doesn't have a little red asterisk in the form to indicate "required", but still throws an error of "This field is required" when you try to save it with no blocks:

```
    assets = StreamField(
        MixedMediaCarouselBlock(),
        help_text='Pick one or more images/videos to place in the sidebar of this article.',
        blank=True
    )
```

Adding `required=False` as an argument to `MixedMediaCarouselBlock()` (which derives from `StreamBlock`) is the only way to make it possible to submit the form with no blocks in the field.

I am also using Wagtail 1.13.1.

**Comment by alexgleason** [comment]
@coredumperror's workaround fixed it for me. Just pass `required=False` into the StreamBlock. Thank you!

I see this is tagged `some-day`. If this is a complicated issue to fix, I think noting the issue in the docs might be worth it. There's a StackOverflow question about it which points to the docs, which aren't correct: https://stackoverflow.com/questions/47336893/how-can-i-make-a-wagtail-streamfield-not-required

I disagree that this is confusing behavior; it's a bug. Here's my reasoning:

* StreamField is a normal Django model field, just like CharField, TextField, etc.
* When a Django model field has `blank=True`, it validates even when the field is empty.
* A StreamField is empty when it contains no blocks.

Therefore, when a SteamField is empty (ie containing no blocks) and has `blank=True`, the form should validate. It doesn't matter if the StreamField could contain blocks with `required=True`.

If the StreamField *does* have blocks added to it, and those blocks are blank with `required=True`, then the form should invalidate *on those specific blocks.*

A StreamField should only be invalid if `blank=False` and the StreamField contains no blocks. The state of the child blocks shouldn't matter to StreamField. All that matters is whether blocks exist or not.

**Comment by gasman** [comment]
While reviewing #5185 I've realised what prompted this misstatement in the documentation: the `StreamField`'s `blank` argument takes precedence over the `StreamBlock`'s `required` flag if the `StreamBlock` definition is passed as a _class_, rather than an instance. For example:

    class CarouselBlock(blocks.StreamBlock):
        image = ImageChooserBlock()
        video = EmbedChooserBlock()

        class Meta:
            required = True

    class BlogPage(Page):
        # CarouselBlock is a class - StreamField's blank=True overrides CarouselBlock's required=True
        carousel = StreamField(CarouselBlock, blank=True)

    class GalleryPage(Page):
        # CarouselBlock() is an instance - StreamField's blank=True does NOT override CarouselBlock's required=True
        carousel = StreamField(CarouselBlock(), blank=True)


Clearly, whichever one of `StreamField.blank` or `StreamBlock.required` is taken as authoritative in any setup, it ought to be the same one responsible for showing/hiding the asterisk as performing the actual validation, and the fact that it isn't is a bug.

From an implementation perspective, in the case of `StreamField(CarouselBlock(), blank=True)` it would be slightly neater if `blank=True` _didn't_ override `CarouselBlock`'s configuration: conventionally a block's behaviour gets fixed at the point where the block definition object (`CarouselBlock()` here) is instantiated. Currently the block definition object is treated as _almost_ immutable (the one exception being the block's `name` property, so that we can write things like `heading = CharBlock()` in a StreamBlock / StructBlock definition and have `heading` become part of the just-instantiated `CharBlock`'s properties) - making the `blank=True` authoritative would mean modifying the block definition after it had just been instantiated.

However, from the point of view of what makes more sense to site implementers, it's probably more natural for the `StreamField`'s `blank=True` to take precedence, as that one is more tightly bound to the current field definition, i.e.: `CarouselBlock()` might appear in lots of definitions, but in this particular case we're asking for a CarouselBlock that permits blank values. (In the case of something like `foo = StreamField(StreamBlock([...], required=True), blank=True)` it's less obvious which one should "win", as both the StreamField and StreamBlock are one-offs - but there's no logical reason to treat a one-shot definition like `StreamBlock([...], required=True)` as different to a subclass-based one like `CarouselBlock()`, so the same rule applies.)

Given that we're already breaking the "block definition objects are immutable" rule for the `name` property, it's hard to see why the `required` flag shouldn't be another exception to that rule (and, when it comes to #5185, following suit with `min_num`, `max_num` and `block_counts` too). For that reason, I'm inclined to go for the latter option, and change the code to work as described in the docs.

**Comment by timmysmalls** [comment]
@allcaps Are you currently working on this?

**Comment by allcaps** [comment]
No, I wanted to, but gave other things priority. You are welcome to pick this up.

**Comment by gasman** [comment]
Fixed in #6718
